### PR TITLE
Analyzer improve ARRAY JOIN with JOIN

### LINKED ISSUE #43036: Table expression (SELECT [dummy, dummy] FROM system.one) AS x ARRAY JOIN __array_join_expression_0 is not registered in planner context
https://s3.amazonaws.com/clickhouse-test-reports/42865/306b7d76e48d806117ed9f527f526a766d4408e9/fuzzer_astfuzzerubsan//report.html

```
2022.11.07 20:44:58.719296 [ 152 ] {79d09063-35f2-4491-a90a-a8e0963daa41} <Fatal> : Logical error: 'Table expression  (SELECT [dummy, dummy] FROM system.one) AS x ARRAY JOIN __array_join_expression_0 is not registered in planner context'.
2022.11.07 20:45:37.709293 [ 405 ] {} <Fatal> BaseDaemon: ########################################
2022.11.07 20:45:37.709331 [ 405 ] {} <Fatal> BaseDaemon: (version 22.11.1.1, build id: 87237ECE1D9A82461B22177CC82AFF435F704F07) (from thread 152) (query_id: 79d09063-35f2-4491-a90a-a8e0963daa41) (query: SELECT * FROM (SELECT [dummy, dummy] AS dummy FROM system.one) AS x ARRAY JOIN dummy INNER JOIN system.one AS y USING (dummy)) Received signal Aborted (6)
2022.11.07 20:45:37.709356 [ 405 ] {} <Fatal> BaseDaemon:
2022.11.07 20:45:37.709405 [ 405 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f47b9d3000b 0x7f47b9d0f859 0x204d8d03 0x204d8f8f 0x163959b9 0x2be8a661 0x2be8aaaa 0x2be96c6a 0x2be916d7 0x2be6c557 0x2be69a24 0x2c320e3a 0x2c31c7ec 0x2d4d6f10 0x2d4f7a1a 0x2e4cc78c 0x2e4ccc7a 0x2e644647 0x2e64212f 0x7f47b9ee7609 0x7f47b9e0c133
2022.11.07 20:45:37.709442 [ 405 ] {} <Fatal> BaseDaemon: 3. gsignal @ 0x7f47b9d3000b in ?
2022.11.07 20:45:37.709467 [ 405 ] {} <Fatal> BaseDaemon: 4. abort @ 0x7f47b9d0f859 in ?
2022.11.07 20:45:37.734594 [ 405 ] {} <Fatal> BaseDaemon: 5. ./build_docker/../src/Common/Exception.cpp:47: DB::abortOnFailedAssertion(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) @ 0x204d8d03 in /workspace/clickhouse
2022.11.07 20:45:37.757021 [ 405 ] {} <Fatal> BaseDaemon: 6. ./build_docker/../src/Common/Exception.cpp:70: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool) @ 0x204d8f8f in /workspace/clickhouse
2022.11.07 20:45:38.851171 [ 405 ] {} <Fatal> BaseDaemon: 7. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>(int, fmt::v8::basic_format_string<char, fmt::v8::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&) @ 0x163959b9 in /workspace/clickhouse
2022.11.07 20:45:38.867876 [ 405 ] {} <Fatal> BaseDaemon: 8. ./build_docker/../src/Planner/PlannerContext.cpp:63: DB::PlannerContext::getTableExpressionDataOrThrow(std::__1::shared_ptr<DB::IQueryTreeNode> const&) const @ 0x2be8a661 in /workspace/clickhouse
2022.11.07 20:45:38.884275 [ 405 ] {} <Fatal> BaseDaemon: 9. ./build_docker/../src/Planner/PlannerContext.cpp:105: DB::PlannerContext::getColumnNodeIdentifierOrThrow(std::__1::shared_ptr<DB::IQueryTreeNode> const&) const @ 0x2be8aaaa in /workspace/clickhouse
2022.11.07 20:45:39.309453 [ 405 ] {} <Fatal> BaseDaemon: 10. ./build_docker/../src/Planner/PlannerJoinTree.cpp:0: DB::(anonymous namespace)::buildQueryPlanForJoinNode(std::__1::shared_ptr<DB::IQueryTreeNode>, DB::SelectQueryInfo&, DB::SelectQueryOptions const&, std::__1::shared_ptr<DB::PlannerContext>&) @ 0x2be96c6a in /workspace/clickhouse
2022.11.07 20:45:39.723134 [ 405 ] {} <Fatal> BaseDaemon: 11. ./build_docker/../src/Planner/PlannerJoinTree.cpp:693: DB::buildQueryPlanForJoinTreeNode(std::__1::shared_ptr<DB::IQueryTreeNode>, DB::SelectQueryInfo&, DB::SelectQueryOptions const&, std::__1::shared_ptr<DB::PlannerContext>&) @ 0x2be916d7 in /workspace/clickhouse
2022.11.07 20:45:39.913387 [ 405 ] {} <Fatal> BaseDaemon: 12. ./build_docker/../src/Planner/Planner.cpp:0: DB::Planner::buildQueryPlanIfNeeded() @ 0x2be6c557 in /workspace/clickhouse
2022.11.07 20:45:39.928326 [ 405 ] {} <Fatal> BaseDaemon: 13. ./build_docker/../src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:94: DB::InterpreterSelectQueryAnalyzer::execute() @ 0x2be69a24 in /workspace/clickhouse
2022.11.07 20:45:39.987099 [ 405 ] {} <Fatal> BaseDaemon: 14. ./build_docker/../src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x2c320e3a in /workspace/clickhouse
2022.11.07 20:45:40.049372 [ 405 ] {} <Fatal> BaseDaemon: 15. ./build_docker/../src/Interpreters/executeQuery.cpp:1073: DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x2c31c7ec in /workspace/clickhouse
2022.11.07 20:45:40.103372 [ 405 ] {} <Fatal> BaseDaemon: 16. ./build_docker/../src/Server/TCPHandler.cpp:374: DB::TCPHandler::runImpl() @ 0x2d4d6f10 in /workspace/clickhouse
2022.11.07 20:45:40.168202 [ 405 ] {} <Fatal> BaseDaemon: 17. ./build_docker/../src/Server/TCPHandler.cpp:1904: DB::TCPHandler::run() @ 0x2d4f7a1a in /workspace/clickhouse
2022.11.07 20:45:40.173590 [ 405 ] {} <Fatal> BaseDaemon: 18. ./build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x2e4cc78c in /workspace/clickhouse
2022.11.07 20:45:40.180819 [ 405 ] {} <Fatal> BaseDaemon: 19.1. inlined from ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:48: std::__1::default_delete<Poco::Net::TCPServerConnection>::operator()[abi:v15003](Poco::Net::TCPServerConnection*) const
2022.11.07 20:45:40.180858 [ 405 ] {} <Fatal> BaseDaemon: 19.2. inlined from ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:305: std::__1::unique_ptr<Poco::Net::TCPServerConnection, std::__1::default_delete<Poco::Net::TCPServerConnection>>::reset[abi:v15003](Poco::Net::TCPServerConnection*)
2022.11.07 20:45:40.180875 [ 405 ] {} <Fatal> BaseDaemon: 19.3. inlined from ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2022.11.07 20:45:40.180889 [ 405 ] {} <Fatal> BaseDaemon: 19. ./build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:116: Poco::Net::TCPServerDispatcher::run() @ 0x2e4ccc7a in /workspace/clickhouse
2022.11.07 20:45:40.189008 [ 405 ] {} <Fatal> BaseDaemon: 20. ./build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:213: Poco::PooledThread::run() @ 0x2e644647 in /workspace/clickhouse
2022.11.07 20:45:40.196266 [ 405 ] {} <Fatal> BaseDaemon: 21.1. inlined from ./build_docker/../contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::assign(Poco::Runnable*)
2022.11.07 20:45:40.196299 [ 405 ] {} <Fatal> BaseDaemon: 21.2. inlined from ./build_docker/../contrib/poco/Foundation/include/Poco/SharedPtr.h:208: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::operator=(Poco::Runnable*)
2022.11.07 20:45:40.196315 [ 405 ] {} <Fatal> BaseDaemon: 21. ./build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:360: Poco::ThreadImpl::runnableEntry(void*) @ 0x2e64212f in /workspace/clickhouse
2022.11.07 20:45:40.196334 [ 405 ] {} <Fatal> BaseDaemon: 22. ? @ 0x7f47b9ee7609 in ?
2022.11.07 20:45:40.196351 [ 405 ] {} <Fatal> BaseDaemon: 23. __clone @ 0x7f47b9e0c133 in ?
2022.11.07 20:45:40.512096 [ 405 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read. (calculated checksum: FD6AB11762394B3975302C7A1686DC99)
```

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes #43036.

### CODE REVIEW

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 1]
```diff
@@ -2868,12 +2887,56 @@ QueryAnalyzer::QueryTreeNodesWithNames QueryAnalyzer::resolveUnqualifiedMatcher(
...
     {
         QueryTreeNodesWithNames matched_expression_nodes_with_column_names;
 
         if (auto * array_join_node = table_expression->as<ArrayJoinNode>())
+        {
+            size_t table_expressions_column_nodes_with_names_stack_size = table_expressions_column_nodes_with_names_stack.size();
+            if (table_expressions_column_nodes_with_names_stack_size < 1)
+                throw Exception(ErrorCodes::LOGICAL_ERROR,
+                    "Expected at least 1 table expressions on stack before ARRAY JOIN processing. Actual {}",
+                    table_expressions_column_nodes_with_names_stack_size);
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:2903]: Seems like only `0` can be here

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 2]
```diff
@@ -2868,12 +2887,56 @@ QueryAnalyzer::QueryTreeNodesWithNames QueryAnalyzer::resolveUnqualifiedMatcher(
...
+            /** Special case with ARRAY JOIN column without alias.
+              *
+              * CREATE TABLE test_table (id UInt64, value String, value_array Array(UInt8)) ENGINE=TinyLog;
+              * SELECT * FROM test_table ARRAY JOIN value_array;
+              *
+              * In matched columns `value_array` must be resolved into array join column.
+              */
+            for (const auto & array_join_column_node : array_join_column_nodes)
+            {
+                if (array_join_column_node->hasAlias())
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:2919]: Didn't understand that.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 3]
```diff
@@ -2136,8 +2136,8 @@ QueryTreeNodePtr QueryAnalyzer::tryResolveIdentifierFromTableExpression(const Id
             if (qualified_identifier_with_removed_part.empty())
                 break;
 
-            if (scope.context->getSettingsRef().prefer_column_name_to_alias
-                && scope.alias_name_to_expression_node.contains(qualified_identifier_with_removed_part[0]))
+            IdentifierLookup bind_to_aliases_identifier_lookup = {qualified_identifier_with_removed_part, IdentifierLookupContext::EXPRESSION};
+            if (tryBindIdentifierToAliases(bind_to_aliases_identifier_lookup, scope))
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp:2140]: This place is bottleneck I will refactor it in next pull request.

File: src/Planner/TableExpressionData.h [hunk 4]
```diff
@@ -164,19 +183,19 @@ class TableExpressionData
...
 private:
-    /// Valid for table, table function, query, union table expression nodes
+    /// Valid for table, table function, query, union, array join table expression nodes
     NamesAndTypesList columns;
 
-    /// Valid for table, table function, query, union table expression nodes
+    /// Valid for table, table function, query, union, array join table expression nodes
     NameSet columns_names;
 
     /// Valid only for table table expression node
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Planner/TableExpressionData.h:192]: Agreed. Will fix in follow up pull request.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp (hunk 2 above)
Comment (kitaisreal) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp:2919]: We can JOIN with ARRAY JOIN columns, and for JOIN ON section we support accessing ARRAY JOIN columns from aliases, but for JOIN with USING section, we try to resolve column USING only from left and right table expression. And to find this column we must add such check here.

**Pull request merged**